04_basic_struct/lds_module.c

```diff
@@ -0,0 +1,183 @@
+// SPDX-License-Identifier: GPL-2.0 OR BSD-3-Clause
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/kernel.h>
+#include <linux/printk.h>
+
+#include <linux/sysfs.h>
+#include <linux/device.h>
+
+#include <linux/types.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+
+MODULE_AUTHOR("Andrii Synenko");
+MODULE_DESCRIPTION("List data module for Linux Kernel ProCamp");
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct list_data {
+	struct list_head my_list;
+	int data;
+};
+
+static LIST_HEAD(data_list_head);
+
+static int list_entry_add(int data, struct list_head *head)
+{
+	struct list_data *entry;
+
+	if (head == NULL) {
+		pr_err("Linked list doesn't exist!\n");
+		return -EINVAL;
+	}
+
+	entry = kmalloc(sizeof(struct list_data), GFP_KERNEL);
+
+	if (entry == NULL)
+		return -ENOMEM;
+
+	entry->data = data;
+
+	list_add(&entry->my_list, head);
+
+	return 0;
+}
+
+static int list_print(struct list_head *head, char *buf)
+{
+	size_t list_size = 0;
+	int res = 0;
+	int cursor = 0;
+
+	struct list_data *entry = NULL;
+
+	if ((head == NULL) || (buf == NULL)) {
+		pr_err("Linked list or output buffer doesn't exist!\n");
+		return -EINVAL;
+	}
+
+	res += sprintf(&buf[cursor], "[head] => ");
+	if (res > 0)
+		cursor += res;
+	else {
+		pr_err("Failed to write output.\n");
+		return -EIO;
+	}
+
+	list_for_each_entry(entry, head, my_list) {
+		res = sprintf(&buf[cursor], "[%d] => ", entry->data);
+		if (res > 0)
+			cursor += res;
+		else {
+			pr_err("Failed to write output.\n");
+			return -EIO;
+		}
+		list_size++;
+	}
+
+	res = sprintf(&buf[cursor], "[head], size = %d node(s)\n",
+		list_size);
+
+	if (res > 0)
+		cursor += res;
+	else {
+		pr_err("Failed to write output.\n");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int list_free(struct list_head *head)
+{
+	struct list_data *entry = NULL;
+	struct list_head *position = NULL,
+					 *tmp = NULL;
+
+	if (head == NULL) {
+		pr_err("Linked list doesn't exist!\n");
+		return -EINVAL;
+	}
+
+	list_for_each_safe(position, tmp, head) {
+		entry = list_entry(position, struct list_data, my_list);
+		list_del(position);
+		kfree(entry);
+	}
+	return 0;
+}
+
+static ssize_t list_show(struct kobject *kobj, struct kobj_attribute *attr,
+					   char *buf)
+{
+	list_print(&data_list_head, buf);
+	return strlen(buf);
+}
+
+static ssize_t list_store(struct kobject *kobj, struct kobj_attribute *attr,
+						const char *buf, size_t count)
+{
+	int res = 0;
+	int tmp_value = 0;
+
+	res = sscanf(buf, "%du", &tmp_value);
```

Comment:
*it should add new **string** to the objects linked list*
Replace `int data;` filed with `const char *data;`
Hint: use `kstrdup` to allocate memory and duplicate string. Free the memory in `list_free()`

**Bonus task:**
In `list_print`, you can not print into `*buf` more than `PAGE_SIZE` bytes. So:
* Add a field to keep the total byte count in list to save time (do not scan the whole list on addition);
* In `list_store` check if there is a room to add a new data. Remove items from head of the list until there is enough space and update the total count field;
* Update the total count field with new data length.

So that you will implement a circular buffer for strings which size is limited with the maximal length to print.
You can also save the total length of deleted rows to print statistics ;-)

Reply:
Done. Bonus task tested with `#define DEBUG`